src/runtime/refiner.ts

```diff
@@ -0,0 +1,140 @@
+/**
+ * @license
+ * Copyright (c) 2019 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Refinement, SchemaPrimitiveType} from './manifest-ast-nodes.js';
+import {Dictionary} from './hot.js';
+import {assert} from '../platform/assert-node.js';
+
+export class Refiner {
+    static refinementString(refinement) : string {
+        if (!refinement) {
+            return '';
+        }
+        return '[' + Refiner.expressionString(refinement.expression) + ']';
+    }
+
+    private static expressionString(expr) : string {
+        if (expr.kind === 'binary-expression-node') {
+            return '(' + Refiner.expressionString(expr.leftExpr) + ' ' + expr.operator + ' ' + Refiner.expressionString(expr.rightExpr) + ')';
+        } else if (expr.kind === 'unary-expression-node') {
+            return '(' + expr.operator + ' ' + expr.expr + ')';
+        }
+        return expr.toString();
+    }
+
+    // tslint:disable-next-line: no-any
+    static refineData(refinement, data) {
+        const result = Refiner.applyRefinement(refinement.expression, data);
+        if (result instanceof Error) {
+            throw result;
+        } else if (typeof result !== 'boolean') {
+            throw new Error(`Refinement expression evaluated to a non-boolean type.\n`);
+        }
+        return result;
+    }
+
+    private static applyRefinement(expr, data) {
+        if (expr.kind === 'binary-expression-node') {
+            const left = Refiner.applyRefinement(expr.leftExpr, data);
+            const right = Refiner.applyRefinement(expr.rightExpr, data);
+            if (left instanceof Error || right instanceof Error) {
+                return Refiner.manageErrors([left, right]);
+            }
+            return Refiner.applyOperator(expr.operator, [left, right]);
+        } else if (expr.kind === 'unary-expression-node') {
+            const curr = Refiner.applyRefinement(expr.expr, data);
+            if (curr instanceof Error) {
+                return Refiner.manageErrors([curr]);
+            }
+            return Refiner.applyOperator(expr.operator, [curr]);
+        }
+        // TODO(ragdev): Update when true, false and string literals are supported
+        // in the refinement expression.
+        if (typeof expr === 'number') {
+            return expr;
+        } else if (typeof expr === 'string') {
```

Comment:
inevitably it's going to end up easier to read/reason about this code if the exprs and ops are instances of classes in their own right. Might be worth making this change sooner rather than later?